xdsl/interpreters/riscv.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Callable, Generic, Iterator, TypeAlias, TypeVar
+
+from xdsl.dialects import riscv
+from xdsl.dialects.builtin import AnyIntegerAttr, IntegerAttr, ModuleOp
+from xdsl.interpreter import (
+    Interpreter,
+    InterpreterFunctions,
+    PythonValues,
+    impl,
+    register_impls,
+)
+from xdsl.interpreters.comparisons import to_signed, to_unsigned
+from xdsl.ir.core import Operation
+from xdsl.utils.exceptions import InterpretationError
+
+_T = TypeVar("_T")
+
+
+def pairs(els: list[_T]) -> Iterator[tuple[_T, _T]]:
+    count = len(els)
+    assert not count % 2
+    for i in range(0, count, 2):
+        yield els[i], els[i + 1]
+
+
+CustomInstructionFn: TypeAlias = Callable[
+    [Interpreter, riscv.CustomAssemblyInstructionOp, PythonValues], PythonValues
+]
+
+
+@dataclass
+class Buffer(Generic[_T]):
+    """
+    Data structure to help simulate pointer offsets into buffers of data in memory.
+    """
+
+    data: list[_T]
+    offset: int = 0
+
+    def __add__(self, offset: int) -> Buffer[_T]:
+        """
+        Aliases the data list, so storing into the offset stores for all other references
+        to the list.
+        """
+        return Buffer(self.data, self.offset + offset // 4)
+
+    def __getitem__(self, key: int) -> _T:
+        return self.data[self.offset + key]
+
+    def __setitem__(self, key: int, value: _T):
+        self.data[self.offset + key] = value
+
+
+@register_impls
+class RiscvFunctions(InterpreterFunctions):
+    module_op: ModuleOp
+    data: dict[str, Any]
+    custom_instructions: dict[str, CustomInstructionFn] = {}
+    bitwidth: int
+
+    def __init__(
+        self,
+        module_op: ModuleOp,
+        *,
+        bitwidth: int = 32,
+        data: dict[str, Any] | None = None,
+        custom_instructions: dict[str, CustomInstructionFn] | None = None,
+    ):
+        super().__init__()
+        self.module_op = module_op
+        self.bitwidth = bitwidth
+        if data is None:
+            data = RiscvFunctions.get_data(module_op)
```

Comment:
This is untested and the only used of `get_data`. Don't you want to test it if you use it?

Reply:
idem for `def pairs(...)` and the `__add__` helper, which AFAIU are meant to be used in `get_data`

Reply:
Ah good point forgot to add tests for this